isthmus/src/main/java/io/substrait/isthmus/expression/FunctionConverter.java

```diff
@@ -69,9 +73,48 @@ public FunctionConverter(
       }
     }
 
+    for (var entry : alm.entries()) {
+      String key = entry.getKey();
+      var func = entry.getValue();
+      for (FunctionMappings.Sig sig : calciteOperators.get(key)) {
+        substraitFuncKeyToSqlOperatorMap.put(func.key(), sig.operator());
+      }
+    }
+
     this.signatures = matcherMap;
   }
 
+  protected Optional<SqlOperator> getSqlOperatorFromSubstraitFunc(
+      Expression.ScalarFunctionInvocation expr) {
+    var resolver = getTypeBasedResolver();
+    if (substraitFuncKeyToSqlOperatorMap.containsKey(expr.declaration().key())) {
+      var operators = substraitFuncKeyToSqlOperatorMap.get(expr.declaration().key());
+      // only one SqlOperator is possible
+      if (operators.size() == 1) {
+        return Optional.of(operators.iterator().next());
+      }
+      // at least 2 operators. Use output type to resolve SqlOperator.
+      String outputTypeStr = expr.outputType().accept(ToTypeString.INSTANCE);
+      var resolvedOperators =
+          operators.stream()
+              .filter(
+                  operator ->
+                      resolver.containsKey(operator)
+                          && resolver.get(operator).types().contains(outputTypeStr))
+              .toList();
+      // only one SqlOperator is possible
+      if (resolvedOperators.size() == 1) {
+        return Optional.of(resolvedOperators.get(0));
+      } else if (resolvedOperators.size() > 1) {
+        throw new RuntimeException(
+            "Found two SqlOperators: " + resolvedOperators + " for ScalarFunction : " + expr);
+      }
+    }
+    return Optional.empty();
+  }
+
+  protected abstract Map<SqlOperator, FunctionMappings.TypeBasedResolver> getTypeBasedResolver();
```

Comment:
Wondering whether we should expose a more concrete function here instead of the internals. (?)

Reply:
Are you suggesting we change `getTypeBasedResolver` to a private function of `FunctionConverter`? Not sure if I understand it.  I changed it to private, and return `OPERATOR_RESOLVER` for both scalar and aggregate operator.

Reply:
Also, after identify convert a substrait function invocation to a SqlOperator, using the above `getTypeBasedResolver`,  previously I added a check if the SqlOperator's returnTypeInference will derive to the same return type, by creating a `RexCallBinding`.  Turns out that approach will not work.  Calcite's SqlOperator's returnTypeInference seems to only work for SqlCallBinding, not for all `RexCallBinding`.   When I worked on interval literal and functions, I found that DateTimeSubstract or DateTimePlus operator assumes 3 arguments for SqlCall, and 2 arguments for RexCall.   The `returnTypeInference` requires 2-rd argment, which is only applicable to SqlCall.  

As such, I have to remove the newly added checking in `ExpressionRexConverter`.

1. https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/sql/fun/SqlDatetimeSubtractionOperator.java#L37-L38
2. https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/sql/fun/SqlDatetimeSubtractionOperator.java#L50